### PR TITLE
services/horizon/ingest: added 'horizon_ingest_errors_total' metric key

### PR DESCRIPTION
**Pull request by sreuland**
<details>
  <summary>PR Checklist</summary>
  
### PR Structure

* [x] This PR has reasonably narrow scope (if not, break it down into smaller PRs).
* [x] This PR avoids mixing refactoring changes with feature changes (split into two PRs
  otherwise).
* [x] This PR's title starts with name of package that is most changed in the PR, ex.
  `services/friendbot`, or `all` or `doc` if the changes are broad or impact many
  packages.

### Thoroughness

* [x] This PR adds tests for the most critical parts of the new functionality or fixes.
* [ ] I've updated any docs ([developer docs](https://developers.stellar.org/api/), `.md`
  files, etc... affected by this change). Take a look in the `docs` folder for a given service,
  like [this one](https://github.com/stellar/go/tree/master/services/horizon/internal/docs).

### Release planning

* [ ] I've updated the relevant CHANGELOG ([here](services/horizon/CHANGELOG.md) for Horizon) if
  needed with deprecations, added features, breaking changes, and DB schema changes.
* [ ] I've decided if this PR requires a new major/minor version according to
  [semver](https://semver.org/), or if it's mainly a patch change. The PR is targeted at the next
  release branch if it's not a patch change.
</details>

### What

added a new metrics counter key 'horizon_ingest_errors_total' which ingest fsm will emit anytime it traps an error response from in ingestion state run. 

### Why

to configure prometheus alert against the new metric for early visibility on when an ingestion halt starts to form - https://github.com/stellar/prometheus/pull/243

Closes: #5256

### CODE REVIEW

**Comment by tamirms** [comment]
We log ingestion errors here:

https://github.com/stellar/go/blob/master/services/horizon/internal/ingest/main.go#L646

If you search on kibana you can see that these logs occur very rarely. I think it would make sense to have an ingestion error counter metric which is basically triggered every time we emit the log in https://github.com/stellar/go/blob/master/services/horizon/internal/ingest/main.go#L646. The metric definition should mirror the log message, so it should be a counter with 2 labels: current_state and next_state.

We can the create the alert using a very simple heuristic: if ingestion error counter is incremented more than `n` times in the last `m` minutes fire the alert

**Comment by sreuland** [comment]
> I think it would make sense to have an ingestion error counter metric which is basically triggered every time we emit the log in https://github.com/stellar/go/blob/master/services/horizon/internal/ingest/main.go#L646.

ah, thanks for pointer to that, updated - https://github.com/stellar/go/pull/5302/commits/3319801ce7c8370db5564811c431e49732fca702

File: go.mod [hunk 1]
```diff
@@ -49,7 +49,7 @@ require (
 	github.com/spf13/viper v1.17.0
 	github.com/stellar/go-xdr v0.0.0-20231122183749-b53fb00bcac2
 	github.com/stellar/throttled v2.2.3-0.20190823235211-89d75816f59d+incompatible
```
Comment (tamirms) [inline go.mod:51]: are the go mod changes intentional?

File: services/horizon/internal/ingest/fsm.go [hunk 2]
```diff
@@ -405,7 +431,7 @@ func (resumeState) GetState() State {
 	return Resume
 }
 
-func (r resumeState) run(s *system) (transition, error) {
+func (r resumeState) run(s *system) (transitionResult transition, errorResult error) {
```
Comment (tamirms) [inline services/horizon/internal/ingest/fsm.go:434]: can you revert this change given that we don't need the named return values anymore?

File: services/horizon/internal/ingest/main.go [hunk 3]
```diff
@@ -443,6 +447,16 @@ func (s *system) initMetrics() {
 		},
 		[]string{"source", "type"},
 	)
+
+	s.metrics.IngestionErrorCounter = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Namespace: "horizon", Subsystem: "ingest", Name: "errors_total",
+			Help: "Counters of the number of times the live/forward ingestion state machine encountered an error. " +
+				"'current_state' label has the name of the state that error occurred. " +
```
Comment (tamirms) [inline services/horizon/internal/ingest/main.go:455]:
```suggestion
				"'current_state' label has the name of the state where the error occurred. " +
```

**Review by tamirms — APPROVED** [review]
I left a few minor comments but overall looks good!

File: go.mod (hunk 1 above)
Comment (sreuland) [reply go.mod:51]: yes, pulled in a race bug fix on `EventuallyWithT`
